README.md
Copyright (c) 2012-2017, The CryptoNote developers, The Bytecoin developers, The Forknote developers  
Copyright (c) 2017-2018, The BitcoiNote developers

# BitcoiNote Core

BitcoiNote (BTCN) is a cryptocurrency based on CryptoNote.

**P2P Port:** 28388  
**RPC Port:** 28389

## Build

Currently, build is supported out of the box only on Linux (tested with Ubuntu 16.04) and Windows (64-bit only). Mac OS should theoretically work but is untested.

### Linux

Tested only with Ubuntu 16.04.

Dependencies: GCC 4.7.3 or later, CMake 2.8.6 or later, Boost 1.58.

You may download them from:  
[http://gcc.gnu.org/](http://gcc.gnu.org/)  
[http://www.cmake.org/](http://www.cmake.org/)  
[http://www.boost.org/](http://www.boost.org/)  
[https://www.qt.io/](https://www.qt.io/)  
Alternatively, it may be possible to install them using a package manager.

To build, run the following commands in the project's root directory:

```
mkdir build
cd build
cmake -DPORTABLE=1 ..
make
```

The resulting executables can be found in `build/src` directory which will be created. (For Windows, look in subfolder `Release` or `Debug` respectively.)

### Windows 64bit

Dependencies:
```
libboost 1.58 or greater (Tested with 1.65.1)
cmake 2.8.6 or greater (Tested with 3.10.2)
Microsoft Visual Studio 2013 (Tested with VS2015)
Visual C++ Build Tools v120 (Tested with v140)
```

To build, run the following commands in the project's root directory, but make sure to substitute the Visual Studio version for the one that you use, and to replace the VCTargetsPath with one that works for you:

```
mkdir build
cd build
set PATH="c:\Program Files (x86)\Windows Kits\10\bin\10.0.15063.0\x64";%PATH%
set VCTargetsPath=C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\v140
cmake -G "Visual Studio 14 Win64" -DPORTABLE=1 -DCMAKE_BUILD_TYPE=Release ..
msbuild Bitcoinote.sln /m /p:Configuration=Release
```

In your build folder there should now be a new folder named `Release`, with the binaries.
You still need to copy some library files into that folder:

```
C:\Windows\System32\msvcp140.dll
C:\Windows\System32\msvcrt.dll
C:\WINDOWS\system32\ucrtbase.dll
C:\WINDOWS\system32\vcruntime140.dll
```

To compile the debug version, omit the `/p:Configuration=Release` parameter. Then you need to copy the debug dlls instead of the normal ones. These have the same name, only with a `d` at the end (e.g. `Qt5Cored.dll`), and are in the same location, except for `ucrtbased.dll` which you can find in `C:\Program Files (x86)\Windows Kits\10\bin\x64\ucrt\ucrtbased.dll`.

### macOS

Tested only with macOS High Sierra 10.13.6.

Dependencies: `clang 9.1.0` or later, `CMake 3.5.2` or later, `Boost 1.67` or later and `Qt5` (tested with `Qt 5.10.1`).

* Use [Homebrew](https://brew.sh/) to install `cmake` and `boost`.
* Use [qt.io](https://www.qt.io/) to install QtCreator.

To build, open the `CMakeLists.txt` in QtCreator. Also you can turn on `PORTABLE` flag in the project settings tab.

## Usage Notes

The following files are created:

* `Bitcoinoted`: The main coin daemon.
* `miner`: Stand-alone miner tool (requires daemon).
* `simplewallet`: Command-line wallet tool (if started with `--rpc-bind-port 1234` it will expose a simplified JSON RPC wallet interface).
* `walletd`: Wallet daemon (exposes full Payment Gate Service RPC interface on default port `8070`).
* `connection_tool`: Connection tool.

Start them with the parameter `--help` to get usage information.


ReleaseNotes.txt
Release notes 2.0.0

- HARD FORK: Starting after block height 178500, block version 5 is used
- New PoW algorithm: Cryptonight Lite V1 ("cryptonight_lite_v7" in xmr-stak-miner)
- Block size limits tweaked to prevent the situation of stuck transactions: TX submission is now rejected again before reaching a stuck state in the TX pool, max. block cumulative sizes are increased by 100k (after hard fork)
- Checkpoints (100000, 130000, 160000)
- Fix a bug which could lead to wallet crashes
- Compiles on macOS now

Release notes 1.2.1

- Fix bug that would cause the network to become stuck when a fusion transaction was submitted that was too big

Release notes 1.2.0

- HARD FORK: Starting after block height 70913, block version 4 is used
- LWMA difficulty algorithm (by zawy12)
- Emergency difficulty recovery - in case the network is stuck for a long time (despite the new algorithm), it is possible to submit blocks with lower difficulty to kick-start the network again
- Checkpoints (70000, 70195, 70920)
- Fix portable compiling on Linux
- Allow much bigger block size
- Faster blockchain syncing
- Checkpoints are checked against blocks in database on daemon startup
- Various smaller fixes and better error output
- Fix bug that wallet synchronization could get stuck

Release notes 1.1.0

- Support for GCC 6+
- Fix many errors "Tx verification failed" displayed on every transaction
- Add restricted RPC mode ("--restricted-rpc")
- Add CORS support ("--enable-cors <domains>")
- Add default fee parameter for simplewallet ("--DEFAULT_FEE")
- Add keys import and export feature for simplewallet (this now makes it possible to recover wallets converted by the GUI wallet to CLI-compatible ones)
- Fix bug in payment gate service regarding empty transaction lists
- Fix off-by-one error in daemon RPC which returned incorrect block for getblockheaderbyheight - IMPORTANT: If you had workarounds in place for this, remove them now!
- Fix security hole regarding double spend attacks with low order point key images
- Add "k_transactions_by_payment_id" RPC method
- Ensure that the RPC server is always returning data type "application/json" when returning JSON (not only at the /json_rpc endpoint) - IMPORTANT: If you were using jQuery you probably used JSON.parse there, remove it now!
- Fix incorrect spelling of "on_getblockhash" RPC command (should be "getblockhash") - old command is still accepted for compatibility reasons
- Fix incorrect spelling of "f_on_transactions_pool_json" RPC command (should be "f_transactions_pool_json")
- Change coin name to BitcoiNote - auto-create symlink for data directory so that both old and new name work (.Bitcoinote and .BitcoiNote)
- Disable start_mining/stop_mining commands in simplewallet because the daemon doesn't support them anyways

Release notes 1.0.2

- Fix (as good as possible) handling of paths with non-ASCII characters

Release notes 1.0.1

- Fix Windows compatibility

Release notes 1.0.0

Initial version


src/CryptoNoteConfig.h
// Copyright (c) 2012-2017, The CryptoNote developers, The Bytecoin developers
//
// This file is part of Bytecoin.
//
// Bytecoin is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Bytecoin is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with Bytecoin.  If not, see <http://www.gnu.org/licenses/>.

#pragma once

#include <cstddef>
#include <cstdint>
#include <limits>

namespace CryptoNote {
namespace parameters {

const uint32_t HF_APR18_HEIGHT                               = 70912; // offset by one! actually 70913
const uint32_t HF_SEP18_HEIGHT                               = 178500; 

const uint32_t CRYPTONOTE_MAX_BLOCK_NUMBER                   = 500000000;
const size_t   CRYPTONOTE_MAX_BLOCK_BLOB_SIZE                = 500000000;
const size_t   CRYPTONOTE_MAX_TX_SIZE                        = 1000000000;
const uint64_t CRYPTONOTE_PUBLIC_ADDRESS_BASE58_PREFIX       = 125;
const uint32_t CRYPTONOTE_MINED_MONEY_UNLOCK_WINDOW         = 100;
const uint64_t CRYPTONOTE_BLOCK_FUTURE_TIME_LIMIT            = 500;

const size_t   BLOCKCHAIN_TIMESTAMP_CHECK_WINDOW             = 60;

// MONEY_SUPPLY - total number coins to be generated
const uint64_t MONEY_SUPPLY                                  = UINT64_C(2100000000000000);
const unsigned EMISSION_SPEED_FACTOR                         = 21;
const uint64_t GENESIS_BLOCK_REWARD                          = UINT64_C(42);
const uint64_t SECOND_BLOCK_REWARD                           = UINT64_C(840000000000000);
static_assert(EMISSION_SPEED_FACTOR <= 8 * sizeof(uint64_t), "Bad EMISSION_SPEED_FACTOR");

const size_t   CRYPTONOTE_REWARD_BLOCKS_WINDOW               = 100;
const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE     = 128000; //size of block (bytes) after which reward for block calculated using block size
const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V4  = 350000; //size of block (bytes) after which reward for block calculated using block size
const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V3  = 20000; //size of block (bytes) after which reward for block calculated using block size
const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V2  = 20000;
const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V1  = 10000;
const size_t   CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_CURRENT = CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE;
const size_t   CRYPTONOTE_COINBASE_BLOB_RESERVED_SIZE        = 600;
const size_t   CRYPTONOTE_DISPLAY_DECIMAL_POINT              = 8;
const uint64_t MINIMUM_FEE                                   = UINT64_C(100000);
const uint64_t DEFAULT_DUST_THRESHOLD                        = UINT64_C(100000);

const uint64_t DIFFICULTY_TARGET                             = 120; // seconds
const uint64_t EXPECTED_NUMBER_OF_BLOCKS_PER_DAY             = 24 * 60 * 60 / DIFFICULTY_TARGET;
const size_t   LEGACY_DIFFICULTY_WINDOW                             = EXPECTED_NUMBER_OF_BLOCKS_PER_DAY; // blocks
const size_t   LEGACY_DIFFICULTY_CUT                                = 60;  // timestamps to cut after sorting
const size_t   LEGACY_DIFFICULTY_LAG                                = 15;  // !!!
const size_t   LWMA_DIFFICULTY_WINDOW                             = 60; // blocks
const size_t   LWMA_DIFFICULTY_CUT                                = 60;  // timestamps to cut after sorting - not used atm
const size_t   LWMA_DIFFICULTY_LAG                                = 0;  // !!!
static_assert(2 * LEGACY_DIFFICULTY_CUT <= LEGACY_DIFFICULTY_WINDOW - 2, "Bad LEGACY_DIFFICULTY_WINDOW or DIFFICULTY_CUT");

const uint32_t MAX_BLOCK_SIZE_INITIAL_SWITCH_BLOCK_HEIGHT    = HF_SEP18_HEIGHT; // Block version 5: MAX_BLOCK_SIZE_INITIAL, previously PREVIOUS_MAX_BLOCK_SIZE_INITIAL
const size_t   PREVIOUS_MAX_BLOCK_SIZE_INITIAL               = 20 * 1024;
const size_t   MAX_BLOCK_SIZE_INITIAL                        = 120 * 1024;
const uint64_t MAX_BLOCK_SIZE_GROWTH_SPEED_NUMERATOR         = 100 * 1024;
const uint64_t MAX_BLOCK_SIZE_GROWTH_SPEED_DENOMINATOR       = 365 * 24 * 60 * 60 / DIFFICULTY_TARGET;

const uint64_t CRYPTONOTE_LOCKED_TX_ALLOWED_DELTA_BLOCKS     = 1;
const uint64_t CRYPTONOTE_LOCKED_TX_ALLOWED_DELTA_SECONDS    = DIFFICULTY_TARGET * CRYPTONOTE_LOCKED_TX_ALLOWED_DELTA_BLOCKS;

const uint64_t CRYPTONOTE_MEMPOOL_TX_LIVETIME                = 60 * 60 * 24;     //seconds, one day
const uint64_t CRYPTONOTE_MEMPOOL_TX_FROM_ALT_BLOCK_LIVETIME = 60 * 60 * 24 * 7; //seconds, one week
const uint64_t CRYPTONOTE_NUMBER_OF_PERIODS_TO_FORGET_TX_DELETED_FROM_POOL = 7;  // CRYPTONOTE_NUMBER_OF_PERIODS_TO_FORGET_TX_DELETED_FROM_POOL * CRYPTONOTE_MEMPOOL_TX_LIVETIME = time to forget tx

const size_t   FUSION_TX_MAX_SIZE                            = MAX_BLOCK_SIZE_INITIAL * 75 / 100;  // was: CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_CURRENT * 30 / 100;
const size_t   FUSION_TX_MIN_INPUT_COUNT                     = 12;
const size_t   FUSION_TX_MIN_IN_OUT_COUNT_RATIO              = 4;

const uint32_t KEY_IMAGE_CHECKING_BLOCK_INDEX                = 0;
const uint32_t UPGRADE_HEIGHT_V2                                = 1;
const uint32_t UPGRADE_HEIGHT_V3                                = 2;
const uint32_t UPGRADE_HEIGHT_V4                                = HF_APR18_HEIGHT;
const uint32_t UPGRADE_HEIGHT_V5                                = HF_SEP18_HEIGHT;
const unsigned UPGRADE_VOTING_THRESHOLD                      = 90;               // percent
const uint32_t UPGRADE_VOTING_WINDOW                         = EXPECTED_NUMBER_OF_BLOCKS_PER_DAY;  // blocks
const uint32_t UPGRADE_WINDOW                                = EXPECTED_NUMBER_OF_BLOCKS_PER_DAY;  // blocks
static_assert(0 < UPGRADE_VOTING_THRESHOLD && UPGRADE_VOTING_THRESHOLD <= 100, "Bad UPGRADE_VOTING_THRESHOLD");
static_assert(UPGRADE_VOTING_WINDOW > 1, "Bad UPGRADE_VOTING_WINDOW");

const char     CRYPTONOTE_BLOCKS_FILENAME[]                  = "blocks.bin";
const char     CRYPTONOTE_BLOCKINDEXES_FILENAME[]            = "blockindexes.bin";
const char     CRYPTONOTE_POOLDATA_FILENAME[]                = "poolstate.bin";
const char     P2P_NET_DATA_FILENAME[]                       = "p2pstate.bin";
const char     MINER_CONFIG_FILE_NAME[]                      = "miner_conf.json";
} // parameters

const char     CRYPTONOTE_NAME[]                             = "BitcoiNote";

const uint8_t  TRANSACTION_VERSION_1                         =  1;
const uint8_t  TRANSACTION_VERSION_2                         =  2;
const uint8_t  CURRENT_TRANSACTION_VERSION                   =  TRANSACTION_VERSION_1;
const uint8_t  BLOCK_MAJOR_VERSION_1                         =  1;
const uint8_t  BLOCK_MAJOR_VERSION_2                         =  2;
const uint8_t  BLOCK_MAJOR_VERSION_3                         =  3;
const uint8_t  BLOCK_MAJOR_VERSION_4                         =  4;
const uint8_t  BLOCK_MAJOR_VERSION_5                         =  5;
const uint8_t  BLOCK_MINOR_VERSION_0                         =  0;
const uint8_t  BLOCK_MINOR_VERSION_1                         =  1;

const size_t   BLOCKS_IDS_SYNCHRONIZING_DEFAULT_COUNT        =  20000;  //by default, blocks ids count in synchronizing
const size_t   BLOCKS_SYNCHRONIZING_DEFAULT_COUNT            =  200;    //by default, blocks count in blocks downloading
const size_t   COMMAND_RPC_GET_BLOCKS_FAST_MAX_COUNT         =  2000;

const int      P2P_DEFAULT_PORT                              =  28388;
const int      RPC_DEFAULT_PORT                              =  28389;

const size_t   P2P_LOCAL_WHITE_PEERLIST_LIMIT                =  1000;
const size_t   P2P_LOCAL_GRAY_PEERLIST_LIMIT                 =  5000;

const size_t   P2P_CONNECTION_MAX_WRITE_BUFFER_SIZE          = 128 * 1024 * 1024; // 128 MB
const uint32_t P2P_DEFAULT_CONNECTIONS_COUNT                 = 12;
const size_t   P2P_DEFAULT_WHITELIST_CONNECTIONS_PERCENT     = 70;
const uint32_t P2P_DEFAULT_HANDSHAKE_INTERVAL                = 60;            // seconds
const uint32_t P2P_DEFAULT_PACKET_MAX_SIZE                   = 100000000;      // 100000000 bytes maximum packet size
const uint32_t P2P_DEFAULT_PEERS_IN_HANDSHAKE                = 250;
const uint32_t P2P_DEFAULT_CONNECTION_TIMEOUT                = 5000;          // 5 seconds
const uint32_t P2P_DEFAULT_PING_CONNECTION_TIMEOUT           = 2000;          // 2 seconds
const uint64_t P2P_DEFAULT_INVOKE_TIMEOUT                    = 60 * 2 * 1000; // 2 minutes
const size_t   P2P_DEFAULT_HANDSHAKE_INVOKE_TIMEOUT          = 5000;          // 5 seconds
const char     P2P_STAT_TRUSTED_PUB_KEY[]                    = "";

const char* const SEED_NODES[] = { "s1.nodes.bitcoinote.org:28388", "s2.nodes.bitcoinote.org:28388" };


struct CheckpointData {
  uint32_t index;
  const char* blockId;
};

const std::initializer_list<CheckpointData> CHECKPOINTS = {
  {70000, "adfbc8ac4f5ae09404a6fbde4a8d06052f9f34d6864a758c54b2f7bc11a5a91d"},
  {70195, "e24699feac664e4c9c75060e299c23b2913f39bcafa9557d79c9601a0084b211"},
  {70920, "a3316d86b15ebe20c9a8bb206dc823f6312d99ff81d4a9b6aa8a8a039b434004"},
  {100000, "1fab25b0991d6584c1f2d519a9df9a4ad1cd8f9fd2fc4ef95647573be2d9adbc"},
  {130000, "88e551f18e24aee324479bfacf8f4a3448706183562447d2b8c523876301cc63"},
  {160000, "2206d1346a32d7d354ea806c3c66ce32d44b999fd05f539672e137c71ae42b68"}
};

} // CryptoNote

#define ALLOW_DEBUG_COMMANDS



